### PR TITLE
Several bug fixes

### LINKED ISSUE #2351: Reconnect button doesn't work (neither in CM nor in status bar)

### LINKED ISSUE #2352: Various remote status don't agree as to where I'm connected
build 2016-09-19.1

DDHUVALO073 is my local (client) machine
DDHUVALO083 is the remote machine

I started VS, R interactive wasn't open.  I opened Workspaces, updated my remote machine name/server, hit save, then connect.

R session says I'm connected to local machine (DDHUVALO073).  Status bar says I'm connected to local as well.  Connection manager and REPL thinks I'm connected to server (DDHUVALO083).

image

### LINKED ISSUE #2354: Connection progress doesn't dismiss when it can't connect
The progress bar dialog stays there forever, despite the connection error in REPL.

image

### LINKED ISSUE #2362: Failed connection should have different message when there is no fallback
1. Connect to a remote machine
1. Close VS
1. Open VS
1. Open Connection Manager. There should be no active connection
1. Try to connect to any connection that will fail

```
Connecting to the new R Workspace failed, restoring connection to 
'' : http://localhost/
```

### LINKED ISSUE #2372: Don't restore connection when new connection failed
- Connecting to remote machine make take time
- User may not necessarily want this
-  It is confusing: we restore local and if user missed the message, repl shows R header and user may have an impression that switch succeeded.

Just output:

'Connection failed. Open Workspaces window and select local R or create another remote connection'.

### PR DESCRIPTION
**Pull request by AlexanderSher**
- Fix #2372: Don't restore connection when new connection failed
- Fix #2362: Failed connection should have different message when there is no fallback
- Fix #2354: Connection progress doesn't dismiss when it can't connect
- Fix #2352: Various remote status don't agree as to where I'm connected
- Fix #2351: Reconnect button doesn't work (neither in CM nor in status bar)

### CODE REVIEW

File: src/Host/Client/Impl/Session/RSessionProvider.cs [hunk 1]
```diff
@@ -172,31 +187,22 @@ public class RSessionProvider : IRSessionProvider {
...
-                    }
-                }
-
-                if (!switchingFromNull) {
-                    _callback.WriteConsole(Resources.RSessionProvider_SwitchingRWorkspaceCompleted);
-                }
+                await SwitchBrokerAsync(cancellationToken, oldBroker);
+                oldBroker.Dispose();
                 PrintBrokerInformation();
+            } catch(Exception) {
```
Comment (int19h) [inline src/Host/Client/Impl/Session/RSessionProvider.cs]: Do we really want to swallow absolutely everything here?

File: src/Host/Client/Impl/Session/RSessionProvider.cs [hunk 2]
```diff
@@ -206,37 +212,109 @@ public class RSessionProvider : IRSessionProvider {
...
+                var transactions = startTransactionTasks.Select(t => t.Result).ToList();
+
+                _callback.WriteConsole(Resources.RSessionProvider_StartConnectingToWorkspaceFormat.FormatInvariant(transactions.Count));
+                await Task.WhenAll(transactions.Select(t => ConnectToNewBrokerAsync(t, cts)));
+
+                _callback.WriteConsole(Resources.RSessionProvider_RestartingSessionsFormat.FormatInvariant(transactions.Count));
+                OnBrokerDisconnected();
+                await Task.WhenAll(transactions.Select(t => CompleteSwitchingBrokerAsync(t, oldBroker, cts)));
+            } finally {
                 foreach (var task in startTransactionTasks.Where(t => t.Status == TaskStatus.RanToCompletion)) {
```
Comment (int19h) [inline src/Host/Client/Impl/Session/RSessionProvider.cs]: This seems like a potential race against task completion. Maybe drop `Where`, and in loop body, do `Dispose` inside `ContinueWith`?

File: src/Host/Client/Impl/Session/RSessionProvider.cs [hunk 3]
```diff
@@ -206,37 +212,109 @@ public class RSessionProvider : IRSessionProvider {
...
+                // All sessions should participate in reconnect. If any of it didn't start, cancel the rest.
+                var startTransactionTasks = sessions.Select(s => s.StartReconnectingAsync(cts.Token)).ToList();
+
+                try {
+                    await Task.WhenAll(startTransactionTasks);
+                    var transactions = startTransactionTasks.Select(t => t.Result).ToList();
+
+                    await Task.WhenAll(transactions.Select(s => ReconnectSessionAsync(s, cts)));
+                } finally {
+                    foreach (var task in startTransactionTasks.Where(t => t.Status == TaskStatus.RanToCompletion)) {
```
Comment (int19h) [inline src/Host/Client/Impl/Session/RSessionProvider.cs]: And same thing here.

File: src/Host/Client/Impl/Session/RSessionProvider.cs [hunk 4]
```diff
@@ -206,37 +212,109 @@ public class RSessionProvider : IRSessionProvider {
...
             // All sessions should participate in switch. If any of it didn't start, cancel the rest.
-            var startTransactionTasks = sessions.Select(s => s.StartSwitchingBrokerAsync(cancellationToken)).ToList();
+            var startTransactionTasks = sessions.Select(s => s.StartSwitchingBrokerAsync(cts.Token)).ToList();
+
             try {
                 await Task.WhenAll(startTransactionTasks);
-            } catch (OperationCanceledException) {
+                var transactions = startTransactionTasks.Select(t => t.Result).ToList();
+
+                _callback.WriteConsole(Resources.RSessionProvider_StartConnectingToWorkspaceFormat.FormatInvariant(transactions.Count));
```
Comment (MikhailArkhipov) [inline src/Host/Client/Impl/Session/RSessionProvider.cs]: You can remove 'starting' and success messages. There is a progress dialog and we print session info (which will be happening on success). Less clutter in REPL

File: src/Host/Client/Impl/Session/RSessionProvider.cs (hunk 2 above)
Comment (AlexanderSher) [inline src/Host/Client/Impl/Session/RSessionProvider.cs]: There is no race.
After tasks are acquired, we do `await Task.WhenAll(startTransactionTasks);`, which will ensure that all tasks are completed.

---
In reply to: [80102485](https://github.com/Microsoft/RTVS/pull/2384#discussion_r80102485) [](ancestors = 80102485)

File: src/Host/Client/Impl/Session/RSessionProvider.cs (hunk 1 above)
Comment (AlexanderSher) [inline src/Host/Client/Impl/Session/RSessionProvider.cs]: Yeah, probably I should should rethrow exceptions that aren't OCE

---
In reply to: [80101432](https://github.com/Microsoft/RTVS/pull/2384#discussion_r80101432) [](ancestors = 80101432)

File: src/Host/Client/Impl/Session/RSessionProvider.cs (hunk 4 above)
Comment (AlexanderSher) [inline src/Host/Client/Impl/Session/RSessionProvider.cs]: Ok

---
In reply to: [80108224](https://github.com/Microsoft/RTVS/pull/2384#discussion_r80108224) [](ancestors = 80108224)

**Comment by MikhailArkhipov** [comment]
I have more changes around here, namely reporting actual errors rather than 500 and 404, I will merge and submit PR.

**Comment by MikhailArkhipov** [comment]
👍

File: src/Host/Client/Impl/Session/RSessionProvider.cs (hunk 2 above)
Comment (int19h) [inline src/Host/Client/Impl/Session/RSessionProvider.cs]: Then `Where` shouldn't be necessary?

**Comment by int19h** [comment]
👍

**Comment by AlexanderSher** [comment]
Do you also print them to the REPL?

---
In reply to: [248991540](https://github.com/Microsoft/RTVS/pull/2384#issuecomment-248991540) [](ancestors = 248991540)

**Comment by MikhailArkhipov** [comment]
Yes I do. I consolidated error printing in one place since we sometimes print several times.

File: src/Host/Client/Impl/Session/RSessionProvider.cs (hunk 2 above)
Comment (AlexanderSher) [inline src/Host/Client/Impl/Session/RSessionProvider.cs]: Without `Where`, `task.Result` will throw exception on the tasks that have failed.

---
In reply to: [80110049](https://github.com/Microsoft/RTVS/pull/2384#discussion_r80110049) [](ancestors = 80110049)

**Comment by MikhailArkhipov** [comment]
Merge?

**Pull request merged**